content/fr/articles/js/flow/advanced-part-1.md

```diff
@@ -0,0 +1,328 @@
+const baz: Set<number> = new Set();
+baz.add(1);
+
+const p: Promise<number> = Promise.resolve(42);
+```
+
+## La manipulation d'objets
+
+Avez vous remarqué que pour le moment, j'ai tenté de ne pas utiliser d'objets au sein de mes exemples (oui, c'était chiant)? C'est tout simplement car les possibilités de manipulation des types de ceux-ci sont très nombreuses. Je vous propose un exemple fleuve histoire d'y voir plus clair.
+
+Un type objet se définit de la sorte :
+
+```js
+// @flow
+
+type User = {
+  name: string,
+  pseudo: string
+};
+
+const user: User = {
+  name: "Mathieu",
+  pseudo: "zoontek"
+};
+
+// Ce qui n'exclut pas de faire
+user.age = 26; // pas d'erreur
+```
+
+Pour sceller les propriétés de notre objet, on utilise la notation `{||}` :
+
+```js
+// @flow
+
+type User = {|
+  name: string,
+  pseudo: string
+|};
+
+const user: User = {
+  name: "Mathieu",
+  pseudo: "zoontek"
+};
+
+user.age = 26; // erreur! la propriété age n'existe pas
+```
+
+Pour indiquer qu'une propriété de notre objet est immutable, on utilise la notation `+`. Un must-have pour vos reducers Redux.
+
+```js
+// @flow
+
+type User = {
+  +name: string,
+  pseudo: string
+};
+
+const user: User = {
+  name: "Mathieu",
+  pseudo: "zoontek"
+};
+
+user.name = "Jean-Michel"; // erreur! la valeur de name ne peut-être modifiée
+
+const newUser: User = {
+  ...user,
+  name: "Jean-Michel"
+}; // pas d'erreur
+```
+
+> Le spreading c'est vraiment cool! Ça existe aussi pour les types?
+
+Yep.
+
+```js
+// @flow
+
+type Foo = {| a: number, b: string |};
+type Bar = {| a: string |};
+type Baz = {| ...Foo, ...Bar |}; // {| a: string, b: string |}
+```
+
+Un dernier petit trick sur les objets pour la route? Vous pouvez les utiliser comme maps :
+
+```js
+// @flow
+
+const foo: { [key: string]: number } = {
+  bar: 42,
+  baz: "whatever" // erreur! les valeurs des propriétés de cet objet doivent être de type number
+};
+```
+
+## Les unions et intersections de types
+
+Si vous avez remarqué la notation avec `|` plus tôt, vous vous posez sûrement la question de ce que ça représente. Il s'agit d'une union de types : la variable aura une valeur à plusieurs types possibles. Petit conseil : utilisez toujours une union de types litéraux à la place du simple `string` lorsque vous connaissez à l'avance les possibles valeurs de celui-ci.
+
+```js
+// @flow
+
+function sayHelloOrRound(arg: string | number) { // arg peut être une string ou un number
+  return typeof arg === "string" ? `Hello ${arg}` : Math.round(arg);
+} // pas d'erreur
+
+type Color =
+  | "red"
+  | "green"
+  | "blue"
+
+function toHexadecimal(color: Color) {
+  switch (color) {
+    case "red":
+      return "#FF0000";
+    case "green":
+      return "#00FF00";
+    case "blue":
+      return "#0000FF";
+  }
+}
+
+toHexadecimal("red"); // pas d'erreur
+toHexadecimal("green"); // pas d'erreur
+toHexadecimal("blue"); // pas d'erreur
+toHexadecimal("pink"); // erreur! "pink" n'est pas une des possibles valeurs
```

Comment:
La formulation "... n'est pas une des possibles valeurs" me parait très anglaise ("...is not a possible value") étrange. Je trouve "... n'est pas une valeur possible" moins bizarre.

Reply:
"ne fait pas partie des valeurs possibles" serait bien